ground/gcs/src/plugins/pathplanner/flightdatamodel.cpp

```diff
@@ -452,8 +456,118 @@ void FlightDataModel::showErrorDialog(const char* title, const char* message)
     msgBox.setStandardButtons(QMessageBox::Ok);
     msgBox.exec();
     return;
+}
+
+void FlightDataModel::pauseValidation(bool pausing) {
+    valPaused = pausing;
+
+    if (!pausing) {
+        fixupValidationErrors();
+    }
+}
+
+void FlightDataModel::fixupValidationErrors()
+{
+    struct FlightDataModel::NED prevNED;
+
+    // Skip validation when there's a download from firmware in process.
+    if (valPaused) { return; }
+
+    for (int i = 0; i < rowCount(); i++) {
+	PathPlanData * row = dataStorage.at(i);
+
+        bool dirty = false;
+
+        struct FlightDataModel::NED thisNED = getNED(row);
+
+        if (i == 0) {
+            row->mode        = Waypoint::MODE_VECTOR;
+            switch (row->mode) {
+		case Waypoint::MODE_CIRCLELEFT:
+		case Waypoint::MODE_CIRCLERIGHT:
+		    row->mode = Waypoint::MODE_VECTOR;
+
+		    showErrorDialog("Waypoint corrected",
+			    "First waypoint may not be the endpoint of an arc");
+                    dirty=true;
+
+		    break;
+		default:
+		    break;
+	    }
+
+        } else {
+	    double distance = sqrt(pow(thisNED.North - prevNED.North, 2) +
+		pow(thisNED.East - prevNED.East, 2));
+
+	    if (distance > 600) {
```

Comment:
Yes.  We use something called the local tangent plane which treats the earth like it's a flat plane around the home location.  It gets increasingly inaccurate as we get away from home.  So validation rejects overly-long legs right now until that code can catch up (we'll probably want to warn in that case, too, because half a kilometer is an awful long leg).